packages/node-core/src/indexer/poi/poi.service.ts

```diff
@@ -1,55 +1,261 @@
+import {PoiEvent} from '../../events';
+import {getLogger} from '../../logger';
+import {ProofOfIndex, SyncedProofOfIndex} from '../entities/Poi.entity';
 import {StoreCacheService} from '../storeCache';
 import {CachePoiModel} from '../storeCache/cachePoi';
+import {PoiBlock} from './PoiBlock';
 
-const DEFAULT_PARENT_HASH = hexToU8a('0x00');
+const GENESIS_PARENT_HASH = hexToU8a('0x00');
+const logger = getLogger('ProofOfIndex');
+
+const syncingMsg = (start: number, end: number, size: number) =>
+  logger.info(`Synced POI [${start} - ${end}] Poi, total ${size} blocks `);
 
 @Injectable()
 export class PoiService implements OnApplicationShutdown {
   private isShutdown = false;
-  private latestPoiBlockHash?: Uint8Array | null;
-  private poiRepo?: CachePoiModel;
+  private _poiRepo?: CachePoiModel;
+  private _latestSyncedPoi?: ProofOfIndex;
+  private isSyncing = false;
+  private _projectId?: string;
 
-  constructor(private storeCache: StoreCacheService) {}
+  constructor(private storeCache: StoreCacheService, protected eventEmitter: EventEmitter2) {}
 
   onApplicationShutdown(): void {
     this.isShutdown = true;
   }
 
-  async init(): Promise<void> {
-    this.poiRepo = this.storeCache.poi ?? undefined;
-    this.latestPoiBlockHash = await this.getLatestPoiBlockHash();
+  get poiRepo(): CachePoiModel {
+    if (!this._poiRepo) {
+      throw new Error(`No poi repo inited`);
+    }
+    return this._poiRepo;
+  }
+
+  get projectId(): string {
+    if (!this._projectId) {
+      throw new Error(`No project Id inited`);
+    }
+    return this._projectId;
+  }
+
+  get latestSyncedPoi(): SyncedProofOfIndex {
+    if (!this._latestSyncedPoi) {
+      throw new Error('_latestSyncedPoi in Poi service is not inited');
+    }
+    if (this._latestSyncedPoi.parentHash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, parentHash is missing`);
+    }
+    if (this._latestSyncedPoi.hash === undefined) {
+      throw new Error(`Invalid latest synced poi at height ${this._latestSyncedPoi.id}, hash is missing`);
+    }
+    return <SyncedProofOfIndex>this._latestSyncedPoi;
+  }
+
+  async init(schema: string, projectId: string): Promise<void> {
+    this._poiRepo = this.storeCache.poi ?? undefined;
+    this._projectId = projectId;
+    const latestSyncedPoiHeight = await this.storeCache.metadata.find('latestSyncedPoiHeight');
+    if (latestSyncedPoiHeight !== undefined) {
+      const recordedPoi = await this.poiRepo.getPoiById(latestSyncedPoiHeight);
+      if (recordedPoi) {
+        await this.setLatestSyncedPoi(recordedPoi);
+      } else {
+        // TODO, could use sql to find latestSyncedPoiHeight in db
+        throw new Error(`Can not find latestSyncedPoiHeight ${latestSyncedPoiHeight}`);
+      }
+    }
+    await this.migratePoi(schema);
+  }
+
+  // This could use for a subcommand later
+  async migratePoi(schema: string): Promise<void> {
```

Comment:
Does this need to clean up the MMR table? Or file based MMR?

Reply:
It cleans up mmr table if exist when migration in the first time. 

For fileBased mmr, because we have removed relate flags for mmr, we could ask user remove it manually.

Reply:
Maybe we can log that they should clean it up?